ext/opcache/jit/ir/ir_strtab.c

```diff
@@ -0,0 +1,227 @@
+/*
+ * IR - Lightweight JIT Compilation Framework
+ * (String table)
+ * Copyright (C) 2022 Zend by Perforce.
+ * Authors: Dmitry Stogov <dmitry@php.net>
+ */
+
+#include "ir.h"
+#include "ir_private.h"
+
+typedef struct _ir_strtab_bucket {
+	uint32_t    h;
+	uint32_t    len;
+	const char *str;
+	uint32_t    next;
+	ir_ref      val;
+} ir_strtab_bucket;
+
+static uint32_t ir_str_hash(const char *str, size_t len)
+{
+	size_t i;
+	uint32_t h = 5381;
+
+    for (i = 0; i < len; i++) {
+        h = ((h << 5) + h) + *str;
+    }
+    return h | 0x10000000;
+}
+
+static uint32_t ir_strtab_hash_size(uint32_t size)
+{
+	size -= 1;
+	size |= (size >> 1);
+	size |= (size >> 2);
+	size |= (size >> 4);
+	size |= (size >> 8);
+	size |= (size >> 16);
+	return size + 1;
+}
+
+static void ir_strtab_resize(ir_strtab *strtab)
+{
+	uint32_t old_hash_size = (uint32_t)(-(int32_t)strtab->mask);
+	char *old_data = strtab->data;
+	uint32_t size = strtab->size * 2;
+	uint32_t hash_size = ir_strtab_hash_size(size);
+	char *data = ir_mem_malloc(hash_size * sizeof(uint32_t) + size * sizeof(ir_strtab_bucket));
+	ir_strtab_bucket *p;
+	uint32_t pos, i;
+
+	memset(data, -1, hash_size * sizeof(uint32_t));
+	strtab->data = data + (hash_size * sizeof(uint32_t));
+	strtab->mask = (uint32_t)(-(int32_t)hash_size);
+	strtab->size = size;
+
+	memcpy(strtab->data, old_data, strtab->count * sizeof(ir_strtab_bucket));
+	ir_mem_free(old_data - (old_hash_size * sizeof(uint32_t)));
+
+	i = strtab->count;
+	pos = 0;
+	p = (ir_strtab_bucket*)strtab->data;
+	do {
+		uint32_t h = p->h | strtab->mask;
+		p->next = ((uint32_t*)strtab->data)[(int32_t)h];
+		((uint32_t*)strtab->data)[(int32_t)h] = pos;
+		pos += sizeof(ir_strtab_bucket);
+		p++;
+	} while (--i);
+}
+
+static void ir_strtab_grow_buf(ir_strtab *strtab, uint32_t len)
```

Comment:
> Are we using a singular buffer for all strings?

yes.

> Is that a good idea?

In my experience with PHP-7 improvement - yes.

> Large allocations might be hard on the allocator, lead to fragmentation, and require a lot of copying. As I understand it, the strings don't need to be contiguous.

For now I didn't see any problem with reallocations (we reuse Zend MM).
In case this becomes a problem, the implementation may be changed (e.g. using linked lists).